### DISCUSSION TITLE
How to use `useAsyncData()` inside plugin

### DISCUSSION THREAD

**Discussion by davestewart** [nuxt/nuxt#16196] [Questions]
Hello.

> Within your pages, components and plugins you can use useAsyncData to get access to data that resolves asynchronously.

Can someone provide an example of using `useAsyncData()` or `useFetch()` inside a plugin please?

I am already using `useState()` which works nicely, the next step is to use one of these async functions and update the state if possible (maybe it's not).

I have `useAsyncData()` working inside a page, but don't understand how to use it in a plugin.

```js
import { defineNuxtPlugin, useAsyncData } from '#app'

function state () {
  return {
    value: 1,
  }
}

function initState (state) {
  return fetch('/api/value')
    .then(res => res.json())
    .then(value => state.value = value)
}

export default defineNuxtPlugin(function (nuxt) {
  // define state (called in server phase)
  useState('test', state)

  // then, load state (don't get how this would work). await seems to break things here too...
  useAsyncData('test', () => {
    const state = ... // is it possible to reference state in this way?
    initState(state)
  })
})
```

If it's not obvious, I'm trying to replicate Vuex's `state` / `nuxtServerInit()` functionality.

If there are more appropriate ways to do this, I'm all ears, but looking to wrap this up in a plugin (vs a page) to automate a few things.

I've also tried `nuxt.provide('test', fn)` in the plugin but in any page code `inject('test')` results in `[Vue warn]: injection "test" not found.` 

Thanks :)

**Comment by danielroe** [comment]
`useAsyncData` should return a value. You shouldn't rely on any side-effects of running the function. Something like this would probably achieve what you're looking for:

```js
import { defineNuxtPlugin, useFetch } from '#app'

function initState () {
  return {
    value: 1,
  }
}

export default defineNuxtPlugin(async (nuxt) => {
  const { data: state } = await useFetch('/api/value', { default: () => initState() })
  // ...
})
```

If you need do do more than just fetch that endpoint, you can use `useAsyncData` with the same `default` option.

**Reply by davestewart** [reply]
Hey Daniel,

Thanks for coming to my rescue!

Unfortunately, I can't seem to get any combination of this to work.

 Dropping right back down to `useFetch()`:

```js
// /plugins/test.js
import { defineNuxtPlugin, useFetch } from '#app'

const baseURL = 'https://jsonplaceholder.typicode.com'

async function useFetchPlugin (nuxt) {
  const res = await useFetch('/todos/1', { baseURL })
  console.log('useFetchPlugin:', res)
}

export default defineNuxtPlugin(useFetchPlugin)
```

```
Cannot read property '_nuxtOnBeforeMountCbs' of null
  at Module.useAsyncData (file://./.nuxt/dist/server/server.mjs:565:17)  
  at Module.useFetch (file://./.nuxt/dist/server/server.mjs:722:32)  
  at useFetchPlugin (file://./.nuxt/dist/server/server.mjs:5215:43)  
  at file://./.nuxt/dist/server/server.mjs:225:38  
  at callWithNuxt (file://./.nuxt/dist/server/server.mjs:269:13)  
  at applyPlugin (file://./.nuxt/dist/server/server.mjs:225:10)  
  at Module.applyPlugins (file://./.nuxt/dist/server/server.mjs:230:11)  
  at async createNuxtAppServer (file://./.nuxt/dist/server/server.mjs:34:5)  
  at async renderToString (file://./node_modules/vue-bundle-renderer/dist/index.mjs:248:19)  
  at async renderMiddleware (file://./.nuxt/nitro/index.mjs:186:20)
```

**Reply by danielroe** [reply]
Sorry - I unaccountably missed that this was taking place in a plugin - too little coffee! `useAsyncData` and `useFetch` only work in components.

see https://github.com/nuxt/framework/issues/1852#issuecomment-966161569

**Reply by davestewart** [reply]
OK, thanks. Have PR'd docs: https://github.com/nuxt/framework/pull/1925

**Reply by Tofandel** [reply]
Why is useFetch or useRequestFetch not useable in a plugin? Because useRequestFetch is the only way to do a request to the nuxt server bypassing the network stack and it's not useable in a plugin

**Reply by danielroe** [reply]
This is no longer the case. They are usable in plugins.

**Reply by Tofandel** [reply]
Indeed, it seems my problem was because I was using something like this

```js
useAsyncData(async () => await useRequestFetch()(...))
```

While this works in vue components, it doesn't work in client side plugins, it seems context is handled a bit differently server side and client side and context doesn't exist in async functions on the client, after reading the code and the doc of unctx, it seems one is using a transformer in vue components https://github.com/unjs/unctx?tab=readme-ov-file#async-transform but doesn't in plugins and instead uses the default https://github.com/unjs/unctx?tab=readme-ov-file#async-context

So I had to do this


```js
const fetch = useRequestFetch()
useAsyncData(async () => await fetch(...))
```

But given that the transformer or callAsync is available to the server as well, maybe this is a bug and should be fixed no? Because then any `await` in a plugin would lead to hard to debug issues

On the server it runs with async context but not on the client
https://github.com/nuxt/nuxt/blob/eaeda4ee1e04e8f185df86351cca3b35e8ea632d/packages/nuxt/src/app/nuxt.ts#L498-L507

Why is this?

**Comment by davestewart** [comment]
So right now, I'm completely stumped on how to do some kind of async setup within the plugin.

The background to all this, is that this is my [Areas](https://github.com/davestewart/nuxt-areas) module for Vue 3, and I'm working on locating all stores, then initialising any which will have an `initState()` function (WIP):

- https://github.com/davestewart/nuxt-areas-demo/tree/main/app

In the module code, I locate all the stores, then pass this data to the template plugin to set up all global state:

https://github.com/davestewart/nuxt-areas/blob/0007-vue-3-state/src/plugins/vue-3.js

The plugin output is as so:

```js
import { provide } from '/_nuxt/node_modules/.vite/vue.js?v=92d040fd';
import { defineNuxtPlugin, useState } from '/_nuxt/node_modules/nuxt3/dist/app/index.mjs';

import * as _52c9f39c from '/_nuxt/external/user-admin/store.js'
import * as _7f3500dc from '/_nuxt/areas/app/store/index.js'
import * as _123a1543 from '/_nuxt/areas/bar/store.js'
import * as _ca93eb50 from '/_nuxt/areas/baz/store/index.js'
import * as _128bafd0 from '/_nuxt/areas/grouped/qux/store/index.js'
import * as _45b6fa4c from '/_nuxt/areas/grouped/qux/store/test.js'
import * as _552ef765 from '/_nuxt/areas/grouped/nested/quux/store/index.js'
import * as _31df74a5 from '/_nuxt/areas/grouped/nested/quux/store/test.js'


// plugin function
export default defineNuxtPlugin(function (nuxt) {
  initState('admin/users', _52c9f39c)
  initState('app', _7f3500dc)
  initState('bar', _123a1543)
  initState('baz', _ca93eb50)
  initState('grouped/qux', _128bafd0)
  initState('grouped/qux/test', _45b6fa4c)
  initState('nested/quux', _552ef765)
  initState('nested/quux/test', _31df74a5)
  
  console.log('Providing store...')
  nuxt.provide('initStore', function () {
    console.log('Initializing store...')
    return Promise.resolve(['a', 'b', 'c']) // do something here to get the async data
  })
})

// helper function
function initState (namespace, module) {
  const { state, init } = module
  if (module.namespace) {
    namespace = module.namespace
  }
  if (typeof state === 'function') {
    useState(namespace, state)
  }
}
```

Because I can't do anything asynchronous in the plugin, my plan was to «somehow» pass a list of «the modules which have additional async functionality declared» into the app so «they could be called when the root component loads».

My root component looks like this:

```vue
<script setup>
import { inject } from 'vue'

const initStore = inject('initStore') // expecting to get the injected function here
</script>
```

But my injection doesn't work; clearly I don't understand the mechanism here, or are using the wrong API. 

```
[vite] connecting...
client.ts:52 [vite] connected.
areas.js:25 Providing store...
runtime-core.esm-bundler.js:893 <Suspense> is an experimental feature and its API will likely change.

runtime-core.esm-bundler.js:6584 [Vue warn]: injection "initStore" not found. 

  at <Default name="default" > 
  at <AsyncComponentWrapper name="default" > 
  at <NuxtLayout key=0 name=undefined > 
  at <RouterView> 
  at <NuxtPage> 
  at <App> 
  at <Root>
...
```

Would be great if someone can point me at what I am missing, or suggest how I should be doing this.

Thanks

**Reply by crotanite** [reply]
If you use `nuxt.provide` then you'll want to use `useNuxtApp().$initStore`.
The inject function is if you do it via `nuxt.vueApp.provide`.

**Reply by davestewart** [reply]
Thanks Paige. Is any of that documented, do you know?

Also, what is `nuxt.provide()` if it's the wrong one to be using?

**Reply by crotanite** [reply]
You can find some documentation for it here: https://v3.nuxtjs.org/docs/usage/nuxt-app

I'm honestly not entirely sure what the difference is between the two, or which is best to use. The documentation seems to recommend doing it via Nuxt over Vue inject.

**Reply by davestewart** [reply]
Thanks, that's helpful.

So it seems the pairing is as so:

Nuxt:
```js
// plugin
export default defineNuxtPlugin(function (nuxt) {
  nuxt.provide('hello', () => 123)
})
```
```js
// component
import { useNuxtApp } from '#app'
const hello = useNuxtApp().$hello
console.log('hello:', hello())
```


Vue:
```js
// plugin
export default defineNuxtPlugin(function (nuxt) {
  nuxt.vueApp.provide('hello', () => 123)
})
```
```js
// component
import { inject } from 'vue'
const hello = inject('hello')
console.log('hello:', hello())
```

I find the Nuxt docs so completely back to front; why I have to go to `nuxtApp` to find out what how to manage the lifecycle, rather than a lifecycle section of the docs to find out about `nuxtApp` is supremely frustrating. It's like the old joke about Irish directions: "Can you tell me how to get to the post office?" "Well you go down there, and if you see the pub, you've gone too far".

Anyway, thanks again.

**Reply by Kapcash** [reply]
If I may clarify 🧐

Both `provide` methods are distincts concepts.
* Vue [provide](https://v3.vuejs.org/guide/component-provide-inject.html) is about injecting some data / functions from a component to its children component in the component tree. You need to use `inject` to declare which provided property you want to use. It's mostly useful for component libraries authors (but not limited to)

* Nuxt provide is about injecting a data / function from a global script (a plugin) to the entire Nuxt app (not just a subtree of the component arborescence but anything that has access to the nuxt instance).

*(These concepts are exactly the same with Vue 2 and Nuxt 2, just a different syntax)*

The Nuxt 3 docs are obviously limited as of now, and a lot of things are just not mentionned yet :/

**Reply by davestewart** [reply]
Thanks @Kapcash - yeah, my general Nuxt knowledge is a fairly sporadic (did I mention the docs already 😉  ?) as I learn bits and bobs as I go along. The gaps are classic "don't know what you don't know" so it's hard to know where to start. You are correct here, that yes, I just want to make some data available at initialisation, rather than inject into a tree (though both will get the job done). As mentioned, this is really a lifecycle issue rather than an access issue. Let me fumble along to the next phase in the exploration and no doubt I'll be back to whinge, ask questions, or report progress! 🙏

**Comment by davestewart** [comment]
OK, have installed the latest Nuxt with the updates from #1931.

This is the latest version of the (rendered) plugin code from [Areas](https://github.com/davestewart/nuxt-areas) for Nuxt 3:

- https://github.com/davestewart/nuxt-areas/blob/0007-vue-3-state/src/plugins/vue-3.js

It outputs this code:

```js
import { provide } from '/_nuxt/node_modules/.vite/vue.js?v=51efec97';// imports
import { defineNuxtPlugin, useState, useAsyncData } from '/_nuxt/node_modules/nuxt3/dist/app/index.mjs'

// store imports
import * as _6cd3b511 from '/_nuxt/external/user-admin/store.js'
import * as _540c4387 from '/_nuxt/areas/app/store/index.js'
import * as _e9279610 from '/_nuxt/areas/bar/store.js'
import * as _2e5cce4d from '/_nuxt/areas/baz/store/index.js'
import * as _b384a276 from '/_nuxt/areas/grouped/qux/store/index.js'
import * as _1b275d76 from '/_nuxt/areas/grouped/qux/store/test.js'
import * as _7381315a from '/_nuxt/areas/grouped/nested/quux/store/index.js'
import * as _1a13a810 from '/_nuxt/areas/grouped/nested/quux/store/test.js'

// store info
const stores = [
  { module: _6cd3b511, namespace: 'admin/users' },
  { module: _540c4387, namespace: 'app' },
  { module: _e9279610, namespace: 'bar' },
  { module: _2e5cce4d, namespace: 'baz' },
  { module: _b384a276, namespace: 'grouped/qux' },
  { module: _1b275d76, namespace: 'grouped/qux/test' },
  { module: _7381315a, namespace: 'nested/quux' },
  { module: _1a13a810, namespace: 'nested/quux/test' }
]

// plugin function
export default defineNuxtPlugin(async function (nuxt) {
  // async
  const asyncStores = []

  // sync
  for (const store of stores) {
    // store contents
    const { namespace, module } = store

    // state functions
    const { state, initState } = module

    // state
    if (typeof state === 'function') {
      // sync state
      const data = useState(namespace, state)

      // async state
      // FIXME: no idea if this is even possible
      if (typeof initState === 'function') {
        // attempt to set state now?
        if (!nuxt.isHydrating) {
          // data.todo = await useAsyncData(namespace, async () => 123)
        }

        // or, attempt to set state later?
        else {
          // this works but makes the network calls in the client
          asyncStores.push({ namespace, initState })
        }
      }
    }
  }

  // final attempt: provide state in app
  // FIXME: not sure *where* to call it in app though - is there a global init-style hook?
  nuxt.provide('initState', function () {
    console.log('Initializing async state')
    return Promise.all(asyncStores.map(async (store) => {
      const state = useState(store.namespace)
      await store.initState(state.value)
      console.log('updated state:', state)
      return state
    }))
  })
})
```

The `useState()` call works fine, and the app loads with the data from the store files..

But I can't seem to get any kind of async data working, whether I use the Nuxt use functions or just do a standard fetch:

```
[Vue warn]: onServerPrefetch is called when there is no active component instance to be associated with. Lifecycle injection APIs can only be used during execution of setup(). If you are using async setup(), make sure to register lifecycle hooks before the first await statement.
nuxt instance unavailable
  at Module.useNuxtApp (file://./.nuxt/dist/server/server.mjs:283:13)  
  at Module.useState (file://./.nuxt/dist/server/server.mjs:699:38)  
  at file://./.nuxt/dist/server/server.mjs:5078:42  
  at async callWithNuxt (file://./.nuxt/dist/server/server.mjs:276:3)  
  at async Module.applyPlugins (file://./.nuxt/dist/server/server.mjs:235:5)  
  at async createNuxtAppServer (file://./.nuxt/dist/server/server.mjs:34:5)  
  at async renderToString (file://./node_modules/vue-bundle-renderer/dist/index.mjs:248:19)  
  at async renderMiddleware (file://./.nuxt/nitro/index.mjs:186:20)  
  at async handle (file://./node_modules/h3/dist/index.mjs:572:19)
```

My fallback is a `nuxt.provide` / `useNuxtApp().$initState()` but this triggers an API call in the client...

I'm just throwing stuff at the wall at the moment because a) I don't know what is going on under the hood and b) my understanding of how the server / client plugin process works has gaps.

Is this worth pursuing right now? Any thoughts @pi0 ?

**Comment by calebsmithdev** [comment]
It's been a bit since you made this post, but were you able to find a solution to this? I'm in a similar boat with trying to make a `useAsyncData` call outside of a component for middleware purposes. I created something similar that works, but it's throwing the same errors.

**Reply by Kapcash** [reply]
`useAsyncData` and `useFetch` composables are only usable from components (setup & hooks functions), not elsewhere. It's by design.
If you need to make http calls, just use `$fetch`, which should be available from the plugin :)
